### ISSUE TITLE
some bug related to setting the shape of sparse matrices

### ISSUE THREAD

**Issue by argriffing** [scipy/scipy#3964]
see https://github.com/scipy/scipy/pull/3957#issuecomment-54399472

**Comment by memeplex** [comment]
For most (all?) formats there is no sensible way to reshape in place in a more efficient manner than calling reshape and replacing the internal representation arrays. Also, the naming convention adopted for set_shape and get_shape is not consistent with the rest of the api (getformat, getcol, getxxx...). Moreover, legacy code depending on assigning to shape is surely broken, so you don't have any backward compatibility issue here (in fact you will be doing a service to that code by signalling an error). Finally, removing the in place reshaping is the easiest solution to implement.

So I see no compelling reason against removing in place reshaping for good.

**Comment by perimosocordiae** [comment]
One real use case I've found is when you want to "expand" a sparse matrix
by adding rows and/or columns of zeros. In many cases, this is a trivial
operation (especially with a COO matrix),  and there should be some
mechanism to do so in-place.

In my ideal world, there would be a warning/error for in-place reshapes
that cause changes to data arrays.

**Comment by memeplex** [comment]
> One real use case I've found is when you want to "expand" a sparse matrix
> by adding rows and/or columns of zeros. In many cases, this is a trivial
> operation (especially with a COO matrix), and there should be some
> mechanism to do so in-place.
>

A related use case I've faced is to build a sparse matrix which shape is
unknown in advance. The initial matrix is "infinite" (in practice, of a
large enough size) in one or both dimensions. After the builder process
finished you do know the real shape of the matrix and would like to resize
the original matrix in a mostly non-destructive way (just changing the
shape parameters would be ideal). But:

1. Often, you still can do it building the i, j, v arrays yourself and then
instantiating a coo_matrix.
2. This is a use case for resize, not for reshape. Resize is easily
implemented for "builder" classes, like dok_matrix.

**Comment by drhagen** [comment]
I believe this was fixed by #7945.

**Comment by sethtroisi** [comment]
PR referenced, code sample produces expected result
```
>>> a = np.array([[1, 2, 3], [4, 5, 6]])
>>> a
array([[1, 2, 3],
       [4, 5, 6]])
>>> from scipy.sparse import lil_matrix
>>> lil = lil_matrix(a)
>>> lil.A
array([[1, 2, 3],
       [4, 5, 6]])
>>> lil.shape
(2, 3)
>>> lil.shape = (3, 2)
>>> lil.shape
(3,2)
# Bug says this would be (2, 3)
```
